sway-parse/src/pattern.rs

```diff
@@ -75,6 +75,19 @@ impl Parse for Pattern {
             return Ok(Pattern::Constructor { path, args });
         }
         if let Some(fields) = Braces::try_parse(parser)? {
+            let _fields: &Punctuated<PatternStructField, CommaToken> = fields.get();
+            let rest_pattern = _fields
```

Comment:
```suggestion
            let fields: &Punctuated<PatternStructField, CommaToken> = fields.get();
            let rest_pattern = fields
```
Typically you only use `_name` to ignore it explicitly.  Is there a reason to not ust use `fields.get().value_separator_pairs...` though?

Reply:
I originally tried to do it like that, but Rust gives me this:

```
error[E0282]: type annotations needed
  --> sway-parse/src/pattern.rs:79:18
   |
78 |               let rest_pattern = fields
   |  ________________________________-
79 | |                 .get()
   | |__________________^^^_- this method call resolves to `&T`
   |                    |
   |                    cannot infer type for type parameter `T`
   |
   = note: type must be known at this point

error[E0609]: no field `value_separator_pairs` on type `&_`
  --> sway-parse/src/pattern.rs:80:18
   |
80 |                 .value_separator_pairs
   |                  ^^^^^^^^^^^^^^^^^^^^^

Some errors have detailed explanations: E0282, E0609.
```

I am still new to Rust so not really sure if there is a better way to fix it.

If I try your suggestion then I also get:

```
error[E0308]: mismatched types
  --> sway-parse/src/pattern.rs:91:47
   |
91 |             return Ok(Pattern::Struct { path, fields });
   |                                               ^^^^^^ expected struct `brackets::Braces`, found reference
   |
   = note: expected struct `brackets::Braces<punctuated::Punctuated<PatternStructField, keywords::CommaToken>>`
           found reference `&punctuated::Punctuated<PatternStructField, keywords::CommaToken>`
```

Should I just rename to something else without an underscore, maybe `inner_fields` or something?

Reply:
So this is one way that works:
```rust
            let rest_pattern = Braces::<Punctuated<PatternStructField, CommaToken>>::get(&fields)
                .value_separator_pairs
                .iter()
                .find(|(p, _)| matches!(p, PatternStructField::Rest { token: _ }));
```
Note that we're using `get()` here as an associated function instead of calling `fields.get()` which seems to need additional type information. `get()` itself is not generic but the impl block where it's defined is. That's why this gets a bit annoying. Maybe @otrho has better ideas.

Reply:
Ahh, right, I didn't notice that you need the original fields later on, so you can't shadow them as I suggested.  And as Mo pointed out the type attribution is needed since `get` is from `Braces<T>`.  I think `Braces::<Punctuated<_, _>>` would still work here, but still a type is needed.

So, really you've done it all correctly, and my only nit was that identifiers beginning with `_` are explicitly ignored by the compiler when reporting unused values.  `rustc` will suggest to prefix unused arguments or patterns with `_` when it finds them 'if this is intentional'.

So yeah, originally I saw this and thought 'why is it unused', but it obviously isn't.  Perhaps renaming `_fields` to `inner_fields` or something similar would be better.